internal/controller/predicates.go

```diff
@@ -0,0 +1,45 @@
+package controller
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// PushPipelineRunCreatedPredicate returns a predicate which filters out all objects except
+// Push PipelineRuns that have just created
+func PushPipelineRunCreatedPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(createEvent event.CreateEvent) bool {
+			return IsPushPipelineRun(createEvent.Object)
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+			return false
+		},
+		GenericFunc: func(genericEvent event.GenericEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+	}
+}
+
+// PushPipelineRunSucceededPredicate returns a predicate which filters out all objects except
+// Push PipelineRuns which have finished and ended successfully
+func PushPipelineRunSucceededPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(createEvent event.CreateEvent) bool {
+			return false
+		},
+		DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+			return false
+		},
+		GenericFunc: func(genericEvent event.GenericEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return IsPushPipelineRun(e.ObjectNew) &&
```

Comment:
we can also check for the annotation in here, so we won't reconcile pipelines we reconciled before.

Reply:
Added